File: rust_engine/src/ttt_layer.rs
use crate::bit_linear::BitLinear;
use ndarray::{Array1, Array2, ArrayView2};

pub struct TTTLayer {
    pub hidden_dim: usize,
    pub inner_lr: f32,
    pub proj_down: BitLinear,
    pub proj_up: BitLinear,
    // Typically the state is per-sequence, but here we might store it if simulating a continuous stream.
    // For TTT, state is usually transient. We'll return it or pass it.
}

impl TTTLayer {
    pub fn new(hidden_dim: usize, inner_lr: f32) -> Self {
        TTTLayer {
            hidden_dim,
            inner_lr,
            proj_down: BitLinear::new(hidden_dim, hidden_dim / 4),
            proj_up: BitLinear::new(hidden_dim / 4, hidden_dim),
        }
    }

    /// Single step update logic
    /// W_state: (D_small, D_small) hidden weights
    /// x_t: (D) input vector
    /// Returns: (Updated W_state, Feature Reconstruction)
    pub fn forward_state_update(
        &self,
        w_state: &Array2<f32>, // Using f32 for state accumulation to be safe for now, can be i32 later
        x_t: &Array1<f32>,
    ) -> (Array2<f32>, Array1<f32>) {
        // 1. Projection
        let mut feat = self.proj_down.forward(x_t); // (D_small)

        // [Optimization] Normalize feat to prevent gradient explosion (Fix B)
        let norm = feat.dot(&feat).sqrt();
        if norm > 1e-6 {
            feat /= norm;
        }

        // 2. Predict Feature from Feature (using W_state)
        // pred_feat = W_state * feat
        // w_state: (D_s, D_s), feat: (D_s)
        let pred_feat = w_state.dot(&feat);

        // 3. Loss = || pred - feat ||^2
        // error = pred - feat
        let error = &pred_feat - &feat;

        // 4. Gradient
        // Grad = error * feat.T  (Outer product)
        // (D_s) * (D_s) -> (D_s, D_s) matrix
        let d_s = feat.len();
        let error_2d = error
            .clone()
            .into_shape((d_s, 1))
            .expect("BUG: error shape mismatch in TTT gradient calculation");
        let feat_2d = feat
            .clone()
            .into_shape((1, d_s))
            .expect("BUG: feat shape mismatch in TTT gradient calculation");
        let grad = error_2d.dot(&feat_2d);

        // 5. Update
        // w_new = w - lr * grad
        let w_new = w_state - (grad * self.inner_lr);

        (w_new, pred_feat)
    }

    pub fn forward_sequence(&self, x: &ArrayView2<f32>) -> Array2<f32> {
        let (seq_len, _dim) = x.dim();
        let d_small = self.hidden_dim / 4;

        let mut w_state = Array2::<f32>::zeros((d_small, d_small));
        let mut outputs = Vec::new();

        for t in 0..seq_len {
            // Extract row t
            let x_t = x.row(t).to_owned(); // Clone for simplicity

            // Update & Predict
            let (w_new, pred_feat) = self.forward_state_update(&w_state, &x_t);
            w_state = w_new;

            // Project Up
            let out_t = self.proj_up.forward(&pred_feat);

            // Residual
            outputs.push(&out_t + &x_t);
        }

        // Stack outputs
        // (This part is a bit verbose in ndarray without helper, assuming flattening)
        let mut result = Array2::<f32>::zeros((seq_len, self.hidden_dim));
        for (i, row) in outputs.iter().enumerate() {
            result.row_mut(i).assign(row);
        }
        result
    }
}

File: legacy/python_proto/ttt_layer.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from bit_linear import BitLinear

class TTTLayer(nn.Module):
    """
    Test-Time Training (TTT) Layer.
    Maintains a hidden weight matrix `W_hidden` that is updated *during the forward pass*
    based on the input sequence.
    """
    def __init__(self, hidden_dim, inner_lr=0.01):
        super().__init__()
        self.hidden_dim = hidden_dim
        self.inner_lr = inner_lr
        
        # The "Learning" Model inside the layer.
        # Ideally this would vary per batch, but for prototype we'll assume batch_size=1
        # or handle state carefully. For simplicity: Batch-independent updates?
        # Standard TTT: State is per-sequence.
        
        # Projection for input to the hidden state dimension
        self.proj_down = BitLinear(hidden_dim, hidden_dim // 4)
        self.proj_up = BitLinear(hidden_dim // 4, hidden_dim)

    def forward_state_update(self, W_state, x_t):
        """
        Perform one step of gradient descent on W_state using x_t.
        Loss Objective: Autoencoder / Reconstruction.
        Goal: Compress x_t into W_state.
        """
        # 1. Prediction (Reconstruction)
        # Using a low-rank approximation: x -> down -> W -> up -> x_recon
        # W_state is (dim, dim) or similar.
        # Let's say W_state maps down_projected -> down_projected
        
        feat = self.proj_down(x_t) # (B, D_small)
        pred_feat = torch.matmul(feat, W_state) # Linear map on features
        
        # 2. Loss: Reconstruction of the feature itself (like simplified predictive coding)
        # L = || pred_feat - feat ||^2
        # Error = pred_feat - feat
        error = pred_feat - feat
        
        # 3. Gradient w.r.t W_state
        # dL/dW = error^T * feat
        grad = torch.matmul(feat.transpose(-2, -1), error)
        
        # 4. Update
        # W_new = W - lr * grad
        # In Bit-TTT, this grad would be quantized to {-1, 0, 1}
        W_state_new = W_state - self.inner_lr * grad
        
        return W_state_new, pred_feat

    def forward(self, x):
        """
        x: (Batch, Seq, HiddenDim)
        """
        B, S, D = x.shape
        D_small = D // 4
        
        # Initial State: W_h (B, D_small, D_small)
        W_state = torch.zeros(B, D_small, D_small, device=x.device)
        
        outputs = []
        
        # Sequential Processing (Causal)
        for t in range(S):
            x_t = x[:, t:t+1, :] # (B, 1, D)
            
            # Update state learning from current input (Testing Time Training)
            # In TTT-Linear, we use W_{t-1} to predict, and update to W_t using x_t.
            # But the output of the layer should benefit from the *updated* state or current?
            # Standard RNN: h_t = f(x_t, h_{t-1}). Output based on h_t.
            
            # Step 1: Update State (Learn)
            # We "learn" the current token to compress it.
            W_state, pred_feat = self.forward_state_update(W_state, x_t)
            
            # Step 2: Generate Output
            # Can use the reconstructed feature or the state itself.
            # Let's project back up.
            out_t = self.proj_up(pred_feat) # (B, 1, D)
            outputs.append(out_t + x_t) # Residual connection
            
        return torch.cat(outputs, dim=1)
